opentech/apply/funds/models.py

```diff
@@ -493,31 +499,71 @@ def current(self):
         return self.exclude(next__isnull=False)
 
 
+def make_permission_check(users):
+    def can_transition(instance, user):
+        if UserPermissions.STAFF in users and user.is_apply_staff:
+            return True
+        if UserPermissions.ADMIN in users and user.is_superuser:
+            return True
+        if UserPermissions.LEAD in users and instance.lead == user:
+            return True
+        if UserPermissions.APPLICANT in users and instance.user == user:
+            return True
+        return False
+
+    return can_transition
+
+
 class AddTransitions(models.base.ModelBase):
     def __new__(cls, name, bases, attrs, **kwargs):
         transition_prefix = 'transition'
         for workflow in WORKFLOWS.values():
             for phase, data in workflow.items():
-                for transition_name, action in data.all_transitions.items():
-                    method = data.transition_methods.get(transition_name)
+                for transition_name, action in data.transitions.items():
+                    method_name = '_'.join([transition_prefix, transition_name, str(data.step), data.stage.name])
+                    permission_name = method_name + '_permission'
+                    permission_func = make_permission_check(action['permissions'])
+
                     # Get the method defined on the parent or default to a NOOP
-                    transition_state = attrs.get(method, lambda self: None)
+                    transition_state = attrs.get(action.get('method'), lambda *args, **kwargs: None)
                     # Provide a neat name for graph viz display
-                    function_name = '_'.join([transition_prefix, slugify(action)])
-                    transition_state.__name__ = function_name
+                    transition_state.__name__ = slugify(action['display'])
+
+                    conditions = [attrs[condition] for condition in action.get('conditions', [])]
                     # Wrap with transition decorator
-                    transition_func = transition(attrs['status'], source=phase, target=transition_name)(transition_state)
+                    transition_func = transition(
+                        attrs['status'],
+                        source=phase,
+                        target=transition_name,
+                        permission=permission_func,
```

Comment:
if I read this correctly, `permission_func` is already true/false, while the `transition()` expects a permission string or a callable?

Reply:
No `permission_func = make_permission_check(action['permissions'])` is returning a callable. The function is just wrapping another function in order to ensure that the `users` set value is stored within the closure. 

```python
def make_permission_check(users):
    def can_transition(instance, user):
        if UserPermissions.STAFF in users and user.is_apply_staff:
            return True
        if UserPermissions.ADMIN in users and user.is_superuser:
            return True
        if UserPermissions.LEAD in users and instance.lead == user:
            return True
        if UserPermissions.APPLICANT in users and instance.user == user:
            return True
        return False

    return can_transition  # A function
```

Reply:
d'oh. re-read `make_permission_check()`. 👍